Ako/ test configuration guideline (#4715)
binary-com/deriv-app: WRITING TESTS.md

# Writing and running tests on Deriv App

> This document is intended to be a guideline for writing and running tests on deriv app.

Reviewer: ```suggestion
This is the ultimate guideline for writing and running tests on *Deriv app*.
```

> Things that this document covers:

Reviewer: ```suggestion
It will cover:
```

> * Brief explanation on test framework implemented on the project.(Jest)

Reviewer: ```suggestion
* Brief explanation on the test framework used in the project (**Jest**)
```

> * We will describe how `react-testing-library` configured. 

Reviewer: ```suggestion
* `react-testing-library` configuration 
```

> * An example of how to write test on using implemented framework

Reviewer: ```suggestion
* Examples
```

>   ## Test Framework explanation 

Reviewer: ```suggestion
  ## Test Framework 
```

>   ​	We are using `Jest` as test runner and [React Testing library](https://testing-library.com/docs/react-testing-library/intro/) as our main test component to have the ultimate needed utilities.

Reviewer: ```suggestion
  ​	We are using [Jest](https://jestjs.io/) as test runner and [React Testing library](https://testing-library.com/docs/react-testing-library/intro/) as the main test component.
```

  

  ### 	Jest 

>   ​		With Jest we can run our tests quickly and even parallel.

Reviewer: ```suggestion
  ​		Jest allows us to run tests quickly and even in parallel.
```

>   ​		All of needed configurations are located in root folder of the app.

Reviewer: ```suggestion
  ​		All the required configurations are in the root folder of the project.
```

  

  #### 	Run Tests

>   ​		To Run tests from root directory simply run:

Reviewer: ```suggestion
  ​		To start the test suite, simply run (from the root directory):
```

  ​			```npm run test:jest```

>   ​		Be informed that this command will run all of the tests in the whole app based on following configuration provided.

Reviewer: ```suggestion
  ​		Be informed that this command runs all the tests in the project.
```

>   ​		Also the mentioned command will generate a coverage directory in the root directory, you can check the current coverage percentage by the generated html.

Reviewer: ```suggestion
  ​		Also it will create a coverage directory in the project root so you can check the current coverage percentage from the generated html.
```

>   #### 	Jest configuration files explanation

Reviewer: ```suggestion
  #### 	Jest configuration
```

  ​	

  * `jest.config.js`
  

>     General configuration of the jest which will use to run the tests from root directory of deriv-app 

Reviewer: ```suggestion
    General configuration of the jest which is used to run the tests from root directory
```

>     1. `collectCoverage`: with this flag jest will collect the coverage report in order to know how much and which files of the code has test coverage.

Reviewer: ```suggestion
    1. `collectCoverage`: A boolean which indicates whether the coverage information should be collected while executing the test.
```

>     2. `collectCoverageFrom`: To indicate which files should and should not test. Generally we exclude all of `node_modules` directories.

Reviewer: ```suggestion
    2. `collectCoverageFrom`: A boolean which indicates which files the coverage info is collected from. Generally we exclude all of `node_modules` directories.
```

>     3. `collectCoverageFrom`: The type of needed coverage report.

Reviewer: Duplicate key? Just like previous one

Author: removed

>     4. `clearMocks`: configuration option to clear mocks automatically before each tests.   

Reviewer: ```suggestion
    4. `clearMocks`: A  boolean to indicate should clear mock calls, instances and results before every test, automatically.   
```

>     5. `projects`: To indicate that which components/directories should test.

Reviewer: ```suggestion
    5. `projects`: An array of paths or glob patterns. Jest will run tests in all of the specified projects at the same time.
```

    6. `transform`: A map from regular expressions to paths to transformers. A transformer is a module that provides a synchronous function for transforming source files.

>     7. `testRegex`: The pattern or patterns Jest uses to detect test files.

Reviewer: ```suggestion
    7. `testRegex`: The pattern/patterns Jest uses to detect test files.
```

>     8. `transformIgnorePatterns`: An array of regexp pattern strings that are matched against all source file paths before transformation. If the file path matches **any** of the patterns, it will not be transformed.

Reviewer: ```suggestion
    8. `transformIgnorePatterns`: An array of regexp patterns that are matched against all source file paths before transformation. If the file path matches **any** of the patterns, it won't be transformed.
```

  

  * `jest.config.base.js`

>     This Configuration file is held in common across all of the individual packages. In other words if you want to just run tests related to one package (eg. trader) this is what that particular package will use as base configuration.

Reviewer: ```suggestion
    This Configuration file exists in all of the individual packages. In other words if you want to just run tests related to one package (e.g. trader) this is what that particular package will use as base configuration.
```

Reviewer 2: It doesn't exist in all of the packages, We just have it in the root directory of the project and we are extending it in the internal packages.

Author: It doesn't exist in all of the packages, We just have it in the root directory of the project and we are extending it in the internal packages

Reviewer: Cool :+1:

>     Also in each component has a `jest.config.js` which extend the `jest.config.base.js` 

Reviewer: ```suggestion
    Also in each component there is a `jest.config.js` file which extends `jest.config.base.js` 
```

>     In this case we have another option to run tests from a specific package inside the app:	

Reviewer: I don't get this part. Is this continued from the last sentence? Or will be continued with the rest? What is the purpose of `:` at the end?

Reviewer 2: Just updated the section to make more clear ;-)

Author: Just updated the section to make it more clear ;-)

    ##### 	Prerequisite

>     ​    	You should install `jest` globally using following command

Reviewer: ```suggestion
    ​    	Install `jest` globally using the following command:
```

    ​			``` npm install jest@[CURRENT_VERSION_OF_JEST_IN_THE_ROOT_PACKAGE.JSON] -g ```
 

##### Run tests from a specific package

>  Navigate to the related directory of package on the command line, then simply run one of the following commands:

Reviewer: ```suggestion
 Navigate to the root directory of the package on the command line, then simply run one of the following commands:
```
